pc_helper_online.py

```diff
@@ -9,35 +12,43 @@ def main():
-    a lot of code which is unrelated to pc_helper
+    a lot of code which is unrelated to pc_helper.
+    As input it getting same folder as bot-ng but processing only two variables which are related to Public Cloud
     """
     log = create_logger("pc_helper_online")
-    settings = {
-        "PUBLIC_CLOUD_TOOLS_IMAGE_QUERY": "https://openqa.suse.de/group_overview/276.json"
-    }
-    log.info("Calling apply_pc_tools_image with :\n%s", settings)
-    apply_pc_tools_image(settings)
-    log.info("Setting after the call to apply_pc_tools_image:\n%s", settings)
-
-    settings = {
-        "PUBLIC_CLOUD_PINT_QUERY": "https://susepubliccloudinfo.suse.com/v1/amazon/images/",
-        "PUBLIC_CLOUD_PINT_NAME": "suse-sles-12-sp5-v[0-9]{8}-hvm-ssd-x86_64",
-        "PUBLIC_CLOUD_PINT_REGION": "us-east-1",
-        "PUBLIC_CLOUD_PINT_FIELD": "id",
-    }
-    log.info("Calling apply_publiccloud_pint_image with :\n%s", settings)
-    apply_publiccloud_pint_image(settings)
-    log.info("Setting after the call to apply_publiccloud_pint_image:\n%s", settings)
-
-    settings = {
-        "PUBLIC_CLOUD_PINT_QUERY": "https://susepubliccloudinfo.suse.com/v1/amazon/images/",
-        "PUBLIC_CLOUD_PINT_NAME": "suse-sles-15-sp2-chost-byos-v[0-9]{8}-hvm-ssd-x86_64",
-        "PUBLIC_CLOUD_PINT_REGION": "us-east-1",
-        "PUBLIC_CLOUD_PINT_FIELD": "id",
-    }
-    log.info("Calling apply_publiccloud_pint_image with :\n%s", settings)
-    apply_publiccloud_pint_image(settings)
-    log.info("Setting after the call to apply_publiccloud_pint_image:\n%s", settings)
+    parser = ArgumentParser(
+        prog="pc_helper_online",
+        description="Dummy code to test functionality related to pc_helper code",
+    )
+    parser.add_argument(
+        "-c",
+        "--configs",
+        type=Path,
+        default=Path("/etc/openqabot"),
+        help="Directory with openqabot configuration metadata",
+    )
+    args = parser.parse_args()
+    loader = YAML(typ="safe")
+    log.info(f"Parsing configuration files from {args.configs}")
+    for p in Path(args.configs).glob("*.yml"):
+        try:
+            data = loader.load(p)
+            log.info(f"Processing {p}")
+            if "settings" in data:
+                settings = data["settings"]
+                if "PUBLIC_CLOUD_TOOLS_IMAGE_QUERY" in settings:
+                    apply_pc_tools_image(settings)
+                    if not settings.get("PUBLIC_CLOUD_TOOLS_IMAGE_BASE", False):
+                        log.error(
+                            f"Failed to get PUBLIC_CLOUD_TOOLS_IMAGE_BASE from {data}"
+                        )
+                if "PUBLIC_CLOUD_PINT_QUERY" in settings:
+                    apply_publiccloud_pint_image(settings)
+                    if not settings.get("PUBLIC_CLOUD_IMAGE_ID", False):
+                        log.error(f"Failed to get PUBLIC_CLOUD_IMAGE_ID from {data}")
+        except Exception as e:
```

Comment:
Since we're using broad exception catching, we could as well avoid checking the keys in the `data` and `settings` dict's.

Reply:
> It's an anti-pattern frowned upon by linters for good reason.

which is ?

> In the case of libraries they usually define a base class for the exceptions they throw

the problem is that you using some libraries which using other libraries . Also undefined combination of different OS errors so at the end you always need several iterations to define all possible error types you want to catch so your loop will keep going in case something will happen. By "iteration" I mean here that you running your program and it get interrupted by new flavor of exception **unexpectedly** ( because you NOT expecting your loop to be interrupted no matter what)  so you creating new commit which add yet-another-handler  . And here comes the fun part **which is doing exactly the same what all other handlers defined before were doing** - log the error and continue .
For me rule is different - **define special error handlers when you have different error handling based on exception type **  E.g. if you got network error switch to offline mode ( if you app has it ) or when you got parsing error of config file try to generate default config . But if your reaction to exception is log the error and continue it does not matter how many handler you have

Reply:
> > It's an anti-pattern frowned upon by linters for good reason.
> 
> which is ?

I should have used plural.  You just mentioned one.  I could add that code with general exception catching can't be reused as functions by code that for some reason needs to catch an exception.

And the problem, again, is not just the general exception catching but the big block within the try/except which makes it harder to find where exactly the exception happened.

This code prints 2:

```
try:
    d = {}
    d[1] = 3
    print(d[2])
except Exception as e:
    print("%s" % e)
```

> the problem is that you using some libraries which using other libraries .

Those libraries should define a subclass of Exception and re-raise most exceptions using it.  But general exception catching makes the problem worse, IMO.

> Also undefined combination of different OS errors so at the end you always need several iterations to define all possible error types you want to catch so your loop will keep going in case something will happen.

I believe that every OS error is covered by `OSError`.

Reply:
```
try:
    d = {}
    d[1] = 3
    print(d[2])
except Exception as e:
    print("%s" % e)
```
I would say this is more about bad exception handling . I can fix it without breaking general exception handling . 

```
import traceback
try:
    d = {}
    d[1] = 3
    print(d[2])
except Exception as e:
    traceback.print_exception(e)
```

can you give me code snippet how this code can be improved by catching more special exception ?

Reply:
> Those libraries should define a subclass of Exception and re-raise most exceptions using it 
we speaking about code which **using** this libraries not how everything should be written in ideal world :)

Reply:
And getting back to my main argument : 

How this is better 

```
for i in i_array:

    try:
        do_something(i)
        do_something2(i)
        do_something3(i)
    except OSError as e:
        log.error("Something happened" % )
    except KeyError as e:
        log.error("Something happened" % )
     except ValueError as e:
       log.error("Something happened" % )
```

than this : 

```
for i in i_array:

    try:
        do_something(i)
        do_something2(i)
        do_something3(i)
    except Exception as e:
        log.error("Something happened" % )
```

? 

or you speaking about "too long blocks in try/catch " so it getting even worth : 

```
```
for i in i_array:

    try:
        do_something(i)
    except OSError as e:
        log.error("Something happened" % )
    except KeyError as e:
        log.error("Something happened" % )
     except ValueError as e:
       log.error("Something happened" % )

    try:
        do_something2(i)
        do_something3(i)
    except OSError as e:
        log.error("Something happened" % )
    except KeyError as e:
        log.error("Something happened" % )
     except ValueError as e:
       log.error("Something happened" % )
```

Reply:
I shouldn't have used that code snippet.

> How this is better
> 
> ```
> for i in i_array:
> 
>     try:
>         do_something(i)
>         do_something2(i)
>         do_something3(i)
>     except OSError as e:
>         log.error("Something happened" % )
>     except KeyError as e:
>         log.error("Something happened" % )
>      except ValueError as e:
>        log.error("Something happened" % )
> ```
> 
> than this :
> 
> ```
> for i in i_array:
> 
>     try:
>         do_something(i)
>         do_something2(i)
>         do_something3(i)
>     except Exception as e:
>         log.error("Something happened" % )
> ```
> 
> ?

If the only thing we do is log the error then `except OSError, KeyError, ValueError as e:` is better.

Reply:
> > Those libraries should define a subclass of Exception and re-raise most exceptions using it
> 
> we speaking about code which **using** this libraries not how everything should be written in ideal world :)

But the YAML library the script is using is well known and you're calling it inside a big try/except block.

Reply:
> If the only thing we do is log the error then except OSError, KeyError, ValueError as e: is better.

I still didn't get how it is better . Only code example which you gave contained problem unrelated to the one which we discussing here . 

So you saying that : 

```
for i in i_array:

    try:
        do_something(i)
        do_something2(i)
        do_something3(i)
    except OSError, KeyError, ValueError as e:
          log.error("Something happened" % )
```
is better than : 

```
for i in i_array:
    try:
        do_something(i)
        do_something2(i)
        do_something3(i)
    except Exception as e:
          log.error("Something happened" % )
```

and I am saying that it is worse because when tomorrow do_something2 will change the code and start throwing also some JSONException or any other NeverSeenBEforeException I will need to come up with extra PR extending this list : 
```
for i in i_array:

    try:
        do_something(i)
        do_something2(i)
        do_something3(i)
    except OSError, KeyError, ValueError, NeverSeenBEforeException as e:
          log.error("Something happened" % )
```

and this is obvious price which you need to pay - when you have exact list you need to keep it up to date with your code . But it is unclear what exactly I will get for this "money" ? Just feeling that I am following some pattern written in some book ? Sorry it is not enough for me . I need something REAL

Reply:
I think we are getting to deep into rabbit hole :) let's move this discussion out of github PR ...